extensions/jms/src/main/java/org/mule/extensions/jms/api/source/JmsMessageListener.java

```diff
@@ -0,0 +1,148 @@
+
+/**
+ * {@link MessageListener} for the {@link JmsListener} to subscribe to a TOPIC or QUEUE receives {@link Message} and
+ * dispatch those messages through the flow.
+ *
+ * @since 4.0
+ */
+public final class JmsMessageListener implements MessageListener {
+
+  private static final Logger LOGGER = getLogger(JmsMessageListener.class);
+  private final JmsSession session;
+  private final SourceCallback<Object, JmsAttributes> sourceCallback;
+  private final JmsListenerLock jmsLock;
+  private final AckMode ackMode;
+  private final String encoding;
+  private final String contentType;
+  private final JmsConfig config;
+  private final JmsSessionManager sessionManager;
+  private final JmsSupport jmsSupport;
+  private final JmsResultFactory resultFactory = new JmsResultFactory();
+
+  /**
+   * Creates a new instance of a {@link JmsMessageListener}
+   *
+   * @param session        the session to create the JMS Consumer
+   * @param config         JMS
+   * @param jmsLock        the lock to use to synchronize the message dispatch
+   * @param sessionManager manager to store the session and ACK ID of each dispatched message
+   * @param sourceCallback callback use to dispatch the {@link Message} to the mule flow
+   * @param jmsSupport     JMS Support that communicates the used specification
+   * @param ackMode        Acknowledgement mode to use to consume the messages
+   * @param encoding       Default encoding if the consumed message doesn't provide one
+   * @param contentType    Default contentType if the consumed message doesn't provide one
+   */
+  JmsMessageListener(JmsSession session, JmsConfig config, JmsListenerLock jmsLock, JmsSessionManager sessionManager,
+                     SourceCallback<Object, JmsAttributes> sourceCallback, JmsSupport jmsSupport, AckMode ackMode,
+                     String encoding, String contentType) {
+    this.session = session;
+    this.sourceCallback = sourceCallback;
+    this.jmsLock = jmsLock;
+    this.ackMode = ackMode;
+    this.encoding = encoding;
+    this.contentType = contentType;
+    this.config = config;
+    this.sessionManager = sessionManager;
+    this.jmsSupport = jmsSupport;
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public void onMessage(Message message) {
+    SourceCallbackContext context = sourceCallback.createContext();
+    evaluateAckAction(message);
+    saveReplyToDestination(message, context);
+    context.addVariable(JMS_LOCK_VAR, jmsLock);
+    context.addVariable(JMS_SESSION_VAR, session);
+    produceMessageResult(message, context, resolveEncoding(message), resolveContentType(message));
+    waitForMessageToBeProcessed(jmsLock);
+  }
+
+  private void evaluateAckAction(Message message) {
+    try {
+      evaluateMessageAck(ackMode, session, message, sessionManager, jmsLock);
+    } catch (JMSException e) {
+      LOGGER.error("An error occurred while processing an incoming message: ", e);
+      sourceCallback.onSourceException(e);
```

Comment:
Why is this not a source exception? I'm communicating that a error occurred trying to process an incoming message.